### DISCUSSION TITLE
How babel merge overrides?

### DISCUSSION THREAD

**Discussion by zdm** [babel/babel#13826] [Q&A] [answered]
Can somebody help me to understand how `babel` merges overrides and why configs below (they are identical on my opinion) are produces different results?
```
// NOT WORK
module.exports = {
    "sourceType": "unambiguous",
    "presets": [["@babel/preset-env", { "useBuiltIns": "usage", "corejs": 3 }]],
    "overrides": [
        {
            "exclude": [/@babel(\/|\\)runtime/, /core-js/],
            "plugins": [["@babel/plugin-transform-runtime", { "version": "^7.8.4" }]],
        },
    ],
};

// WORK
module.exports = {
    "sourceType": "unambiguous",
    "overrides": [
        {
            "include": [/@babel(\/|\\)runtime/],
            "presets": [["@babel/preset-env", { "useBuiltIns": "usage", "corejs": 3 }]],
        },
        {
            "exclude": [/@babel(\/|\\)runtime/, /core-js/],
            "presets": [["@babel/preset-env", { "useBuiltIns": "usage", "corejs": 3 }]],
            "plugins": [["@babel/plugin-transform-runtime", { "version": "^7.8.4" }]],
        },
    ],
};
```

**Comment by nicolo-ribaudo** [comment]
The difference is that in the first config you are also running `@babel/preset-env` on `core-js`, while in the second one you are not.

Some packages break when transpiled; a file somehwere in `core-js` that you are using _might_ be one of those.

**Comment by zdm** [comment]
Do you know, how babel merges presets?
In the working example:
1-st override adds `@babel/preset-env` to the  `babel/runtime` only;
2-nd override adds `@babel/preset-env` and `@babel/plugin-transform-runtime` to the all files, excluding `babel/runtime` and `core-js`.
So it does not add `transform` to the `core-js`, as you wrote.

In the not-working example:
1. It adds `@babel/preset-env` to all files.
2. Adds `@babel/plugin-transform-runtime` also to all files, excluding `babel/runtime` and `core-js`.

Same behavior. Or I am missing something?

**Reply by nicolo-ribaudo** [reply]
Usually when merging plugins and presets, Babel just concatenates the arrays coming from different configs. In case of duplicates, the last specified one "wins" (so if you enable `@babel/preset-env` twice but with different configs, the one in `overrides` replaces the one in the top-level config).

**Reply by zdm** [reply]
So, plugins and presets options are not merged.
Thanks.

**Comment by zdm** [comment, marked as answer]
In another words:
In working example I have:
- for all files: `@babel/preset-env` and `@babel/plugin-transform-runtime`;
- for `babel/runtime`: `@babel/preset-env`;
- for `core-js`: nothing;

In the not0working example:
- for all files: `@babel/preset-env`, `@babel/plugin-transform-runtime`;
- for `babel/runtime`: `@babel/preset-env`;
- for `core-js`: `@babel/preset-env`;

Yes, you were right, Thank you very much.
